odl/util/vectorization.py

```diff
@@ -124,169 +125,240 @@ def out_shape_from_array(arr):
-                if dtype is None:
-                    out = [None] * out_shape[0]  # lazy vectorization
-                else:
-                    out = np.empty(out_shape, dtype=dtype)
-
-            for i, pt in enumerate(x.T):
-                out[i] = func(pt, **kwargs)
-            return np.asarray(out)
-
-        def _vect_wrapper_meshgrid(x, out, ndim, **kwargs):
-            if out is None:
-                out_shape = out_shape_from_meshgrid(x)
-
-                if dtype is None:
-
-                    # Small helper to initialize a nested list
-                    def _nested(shape):
-                        if len(shape) == 0:
-                            return None
-                        else:
-                            return [_nested(shape[1:])
-                                    for _ in range(shape[0])]
-                    out = _nested(out_shape)
-
-                else:
-                    out = np.empty(out_shape, dtype=dtype)
-
-            order = meshgrid_input_order(x)
-            vecs = vecs_from_meshgrid(x, order=order)
-            if ndim == 1:
-                for i, pt in enumerate(vecs[0]):
-                    out[i] = func(pt, **kwargs)
-            elif dtype is None:
-                for i, pt in enumerate(product(*vecs)):
-                    # Worst multi-index implementation ever, but still good
-                    # enough for slow code
-                    index = np.unravel_index(i, out_shape)
-                    sub = out
-                    for j in index[:-1]:
-                        sub = sub[j]
-                    sub[index[-1]] = func(pt, **kwargs)
-            else:
-                for i, pt in enumerate(product(*vecs)):
-                    out.flat[i] = func(pt, **kwargs)
-            return np.asarray(out)
-
-        @wraps(func)
-        def vect_wrapper(x, out=None, **kwargs):
-            # Find out dimension first
-            if isinstance(x, np.ndarray):  # array
-                if x.ndim == 1:
-                    ndim = 1
-                elif x.ndim == 2:
-                    ndim = len(x)
-                else:
-                    raise ValueError('only 1- or 2-dimensional arrays '
-                                     'supported.')
-            else:  # meshgrid or single value
-                try:
-                    ndim = len(x)
-                except TypeError:
-                    ndim = 1
-
-            if is_valid_input_meshgrid(x, ndim):
-                return _vect_wrapper_meshgrid(x, out, ndim, **kwargs)
-            elif is_valid_input_array(x, ndim):
-                return _vect_wrapper_array(x, out, **kwargs)
-            else:
-                try:
-                    return func(x)
-                except Exception as err:
-                    raise_from(
-                        TypeError('invalid vectorized input type.'), err)
-
-        return vect_wrapper
-    return vect_decorator
+    def decorator(func):
+        if not use_kw:
```

Comment:
I'm against keeping anything we dont use, this will simply lead to us having HEAPS of legacy code after a while. If we can remove, please do.